Utils/FaviconFetcher.swift

```diff
@@ -0,0 +1,130 @@
+import Storage
+
+private let queue = dispatch_queue_create("FaviconFetcher", DISPATCH_QUEUE_CONCURRENT)
+
+/* A helper class to find the favicon associated with a url. This will load the page and parse any icons it finds out of it
+ * If that fails, it will attempt to find a favicon.ico in the root host domain
+ */
+class FaviconFetcher : NSObject, NSXMLParserDelegate {
+    static let ExpirationTime = NSTimeInterval(60*60*24*7) // Only check for icons once a week
+
+    private var siteUrl: NSURL // The url we're looking for favicons for
+    private var _favicons = [Favicon]() // An internal cache of favicons found for this url
+
+    class func getForUrl(url: NSURL, profile: Profile, callback: ([Favicon]) -> Void) {
+        dispatch_async(queue) {
+            let f = FaviconFetcher(url: url)
+            f.loadFavicons(profile, callback: callback)
+        }
+    }
+
+    private init(url: NSURL) {
+        siteUrl = url
+    }
+
+    private func loadFavicons(profile: Profile, callback: ([Favicon]) -> Void) {
+        if _favicons.count == 0 {
+            // Initially look for tags in the page
+            loadFromDoc()
+        }
+
+        // If that didn't find anything, look for a favicon.ico for this host
+        if _favicons.count == 0 {
+            loadFromHost()
+        }
+
+        var filledCount = 0
+        for (i, icon) in enumerate(_favicons) {
+            getFavicon(icon, profile: profile) { icon in
+                if let icon = icon {
+                    self._favicons[i] = icon
+                }
+
+                filledCount++
+                if filledCount == self._favicons.count {
+                    self._favicons.sort({ (a, b) -> Bool in
+                        return a.width > b.width
+                    })
+
+                    dispatch_async(dispatch_get_main_queue()) {
+                        callback(self._favicons)
+                    }
+                }
+            }
+        }
+    }
+
+    // Loads favicon.ico on the host domain for this url
+    private func loadFromHost() {
+        var url = NSURL(scheme: siteUrl.scheme!, host: siteUrl.host, path: "/favicon.ico")
+        var request = NSMutableURLRequest(URL: url!)
+        request.HTTPMethod = "HEAD"
+        request.timeoutInterval = 5
+
+        var response: NSURLResponse? = nil;
+        var err = NSErrorPointer()
+        NSURLConnection.sendSynchronousRequest(request, returningResponse: &response, error: err)
```

Comment:
This method and I think most of NSURLConnection is actually deprecated starting in iOS 9 so we'll need to use NRURLSesson/Alamofire for this instead.